src/unified_graphics/diag.py

```diff
@@ -311,19 +94,71 @@ def magnitude(dataset: pd.DataFrame) -> pd.DataFrame:
     )
 
 
-def get_model_run_list(
-    diag_zarr: str,
+def diag_observations(
     model: str,
     system: str,
     domain: str,
     background: str,
     frequency: str,
-    variable: Variable,
-):
-    store = get_store(diag_zarr)
-    path = "/".join([model, system, domain, background, frequency, variable.value])
-    with zarr.open_group(store, mode="r", path=path) as group:
-        return group.group_keys()
+    variable: str,
+    init_time: datetime,
+    loop: str,
+    uri: str,
+    filters: dict = {},
+) -> pd.DataFrame | pd.Series:
+    model_config = "_".join((model, background, system, domain, frequency))
+
+    is_used = filters.pop("is_used", True)
+    parquet_filters = [
+        ("loop", "=", loop),
+        ("initialization_time", "=", init_time),
+    ]
+
+    if isinstance(is_used, bool):
+        parquet_filters.append(("is_used", "=", is_used))
+
+    df = pd.read_parquet(
+        "/".join((uri, model_config, variable)),
+        columns=[
+            "obs_minus_forecast_adjusted",
+            "obs_minus_forecast_unadjusted",
+            "observation",
+            "latitude",
+            "longitude",
+            "is_used",
+        ],
+        filters=parquet_filters,
+    )
+
+    # To apply the filters, we need the vector components in the columns, not
+    # the rows.
+    # FIXME: We should consider changing how we store the vector data so we
+    # don't have to unstack it every time.
+    if "component" in df.index.names:
+        # FIXME: Specifically unstack the component level of the index because
+        # I'm seeing some data where the index is (component, nobs) instead of
+        # (nobs, component)
+        df = df.unstack("component")  # type: ignore
+
+    # Iterate over each filter and apply it
+    for col_name, filter_value in filters.items():
+        arr = np.array(filter_value)
+
+        # Boolean mask for the rows in the data that are within the range
+        # specified by the filter
+        mask = (df[col_name] >= arr.min(axis=0)) & (df[col_name] <= arr.max(axis=0))
+
+        # In the event of a vector variable, we will have a DataFrame mask
+        # instead of a Series, which we need to flatten to a series which
+        # evaluates to True only when every column in the frame is True. If any
+        # column is False, this row should be excluded from the data
+        if len(mask.shape) > 1:
+            mask = mask.all(axis="columns")
```

Comment:
This was very difficult for me to figure out how to do, so those were there as much for me as anything, because coming back even two days later I wouldn’t remember what was happening. (This is something I really dislike about NumPy and Pandas.)